src/main/java/com/openshift/internal/restclient/WatchClient.java

```diff
@@ -52,9 +58,68 @@
 	private IResourceFactory factory;
 	private IClient client;
 	private static WebSocketClient wsClient;
+	private static AtomicReference<Status> status = new AtomicReference<>(Status.Stopped);
+	private static Lock lock = new ReentrantLock();
+	private static Condition isStarted = lock.newCondition();
+	private static long lockTimeout; 
+	
+	private enum Status{
+		Started,
+		Starting,
+		Stopped,
+		Stopping
+	}
 	
 	static {
-		wsClient = newWebSocketClient();
+		try {
+			lockTimeout = Long.parseLong(System.getProperty("com.openshift.restclient.watchlocktimeoutms", "30000")); 
+		}catch(NumberFormatException e) {
+			lockTimeout = 30 * 1000;
+		}
+		try {
+			wsClient = newWebSocketClient();
+			wsClient.addLifeCycleListener(new LifeCycle.Listener() {
+				
+				@Override
+				public void lifeCycleStopping(LifeCycle event) {
+					status.set(Status.Stopping);
+				}
+				
+				@Override
+				public void lifeCycleStopped(LifeCycle event) {
+					status.set(Status.Stopped);
+				}
+				
+				@Override
+				public void lifeCycleStarting(LifeCycle event) {
+					status.set(Status.Starting);
+				}
+				
+				@Override
+				public void lifeCycleStarted(LifeCycle event) {
+					status.set(Status.Started);
+					try{
+						if(lock.tryLock(lockTimeout, TimeUnit.MILLISECONDS)){
+							isStarted.signalAll();
+						}
+					} catch (InterruptedException e) {
+						LOGGER.debug("Exception while trying to get lock",e);
+					}finally {
+						lock.unlock();
+					}
+				}
+				
+				@Override
+				public void lifeCycleFailure(LifeCycle event, Throwable cause) {
+					LOGGER.error("The watchclient failed:", cause);
+					status.set(Status.Stopped);
+				}
+			});
+			
+		}catch(Exception e) {
+			LOGGER.error("Exception initializing watch client",e);
+			throw e;
+		}
```

Comment:
Logging + throwing... wont this end in logging twice since the one user that's launching this will most likely also have a logging statement in his catch block?